### ISSUE TITLE
DetectBeadSubstitutionErrors does not preserve bowtie2 BAM tag values

### ISSUE THREAD

**Issue by jggatter** [broadinstitute/Drop-seq#207]
### Affected tool(s)
DetectBeadSubstitutionErrors

### Affected version(s)
Drop-Seq Tools 2.3.0 Tagged release

### Description 

Hi,

My lab is interested in recovering TCR sequences from . With Drop-Seq Tools, we seek to collapse barcodes via the DetectBeadSynthesisErrors and DetectBeadSubstitutionErrors steps. We substitute alignment via `STAR` with `bowtie2` so that we can get multiple alignments per read to our small reference of V- and J-region variants of _TRA_/_TRB_. After alignment, we did not want to waste time preparing a GTF so we simply used `sed` to append dummy exon and gene tags. The key difference in running the DetectBeadSynthesisErrors and DetectBeadSubstitutionErrors steps is that we do not have many UMIs per barcode, so we set `MIN_UMIS_PER_CELL=1`. What we are finding is that we get some barcode collapsing (~20 mil --> ~19 mil reads) but the values of barcode tags, namely `AS` and `XS`, are wiped to zero. This is only true for the BAM produced by DetectBeadSubstitutionErrors step, specifically.  See below for an example (1 read shown for each):  

`samtools view` after DetectBeadSynthesisErrors. >99% had non-zero value for AS
```
MISEQ:1:1101:1729:13071#CACACTGAGTGTCCTATCCC	16	macfaTRAV21	195	2	4S78M68S	*	0	0	CATAGCTGGATAAACAATCAGGACGGAATACTTTATACATTACAGCTTCTCAGCCGAGTGACTAAGCAACCTACCTCTGTGCGGCCACGGTAAATGGTGCAAGCAACAAGCTAACTGTTGGAGGAGGAACTAGGGTGCTAGTTAAACCAA	#############################AC,9,:,B,ECAFB?B,,9C4+<8,,,F@CDC,@,C,@,C,,6:,,6,@+@F@,7:;,B6,C,<F<,@ECC<<FFEFEGGGGGDE,8E@,GGGGGFGFC-FGGGCGFGGGFGGGGGCCCCC	AS:i:125	XS:i:102	XN:i:0	XM:i:7	XO:i:0	XG:i:0	NM:i:MD:Z:10T0C9T1G27T7C3C14	YT:Z:UU	XC:Z:TTGGTTTAACTA	XM:Z:GCACCCTA	AS:i:0	XS:i:0	CY:Z:CCCCCGGGGGFG	UY:Z:GGFGCGGGGE:Z:GENEA	XF:Z:CODING	GS:Z:+	gf:Z:CODING	gn:Z:GENEA	gs:Z:+
```
`samtools view` after DetectBeadSubstitutionErrors, 100% have 0 for AS (and XS)
```
MISEQ:1:1101:1729:13071#CACACTGAGTGTCCTATCCC	16	macfaTRAV21	195	2	4S78M68S	*	0	0	CATAGCTGGATAAACAATCAGGACGGAATACTTTATACATTACAGCTTCTCAGCCGAGTGACTAAGCAACCTACCTCTGTGCGGCCACGGTAAATGGTGCAAGCAACAAGCTAACTGTTGGAGGAGGAACTAGGGTGCTAGTTAAACCAA	#############################AC,9,:,B,ECAFB?B,,9C4+<8,,,F@CDC,@,C,@,C,,6:,,6,@+@F@,7:;,B6,C,<F<,@ECC<<FFEFEGGGGGDE,8E@,GGGGGFGFC-FGGGCGFGGGFGGGGGCCCCC	XC:Z:TTGGTTTAACTA	MD:Z:10T0C9T1G27T7C3C14	GE:Z:GENEA	XF:Z:CODING	XG:i:0	NM:i:7	XM:Z:GCACCCTA	XN:i:0	XO:i:0	AS:i:0	GS:Z:+	XS:i:0	YT:Z:UU	CY:Z:CCCCCGGGGGFG	UY:Z:GGFGCGGGgf:Z:CODING	gn:Z:GENEA	gs:Z:+
```
Why might the alignment scores and other tag values be wiped to zero? I did not see anything in the source code that stuck out to me.
  
Thanks,
James

#### Steps to reproduce
[bam_history.txt](https://github.com/broadinstitute/Drop-seq/files/5152895/bam_history.txt)  
1. Follow Drop-Seq cookbook with standard arguments
2. After trimming, sort BAM by query name and align it via `bowtie2 -x ref --very-sensitive-local -a -b sorted_trimmed_tagged.bam --preserve-tags -S aligned.sam -p 32`
3. Skip merge step. Convert aligned output SAM to BAM
4. Tag with dummy exons and genes (For our project we don't care about these tags):
```bash
samtools view -H bt2.bam > header_tmp
samtools view bt2.bam | sed 's/$/     GE:Z:GENEA
XF:Z:CODING     GS:Z:+  GF:Z:CODING     GN:Z:GENEA      GS:Z:+/' > tagged_bt2.sam
cat header_tmp tagged_bt2.sam | samtools view -b > tagged_bt2.bam
```
5. DetectBeadSynthesisErrors with min_cell_per_umi=1
6. DetectBeadSubstitutionErrors with min_cell_per_umi=1
7.  Ideally I would stop following the cookbook hereafter and would be able to analyze the bowtie2-tagged and Drop-Seq-Tools-tagged data.

**Comment by jamesnemesh** [comment]
Hi,

I don’t see anything obviously wrong in our code either (The only place the codebase even touches the XS tag is in the default for TagBamWithReadSequenceExtended).  Can you generate a small representative BAM and select a single read (like you did below) where this error occurs so I can test on my side?

-Jim

**Comment by jggatter** [comment]
Hey James,

Thank you for the quick reply! Due to file upload limits, I've emailed you at dropseq@gmail.com the 1-read BAM which was generated after the dummy tagging of exons + genes and before DetectBeadSynthesisErrors. One thing I've noticed is that the BAM right after alignment (pre-tagging) has duplicate AS and XS tags, which is also the case for the BAM I sent you. Looking back, you can also see this in my original post. Generally one of the AS/XS pairs has values of 0 and the other has non-zero values. So I guess it's not a case that Drop-Seq tools is possibly setting tags to zero but rather it's possibly removing the non-zero value set of alignment tags altogether. Eager to hear your thoughts!

Thanks again,
James

**Comment by jamesnemesh** [comment]
Reads having multiple values for the same TAG may be discarding values when read in through HTSJDK.  It's suggested you only have one value per tag, as the behavior of HTSJDK is not specified, but only one value will be retained.

MISEQ:1:1101:1729:13071#CACACTGAGTGTCCTATCCC	16	macfaTRAV21	195	2	4S78M68S	*	0	0	CATAGCTGGATAAACAATCAGGACGGAATACTTTATACATTACAGCTTCTCAGCCGAGTGACTAAGCAACCTACCTCTGTGCGGCCACGGTAAATGGTGCAAGCAACAAGCTAACTGTTGGAGGAGGAACTAGGGTGCTAGTTAAACCAA	#############################AC,9,:,B,ECAFB?B,,9C4+<8,,,F@CDC,@,C,@,C,,6:,,6,@+@F@,7:;,B6,C,<F<,@ECC<<FFEFEGGGGGDE,8E@,GGGGGFGFC-FGGGCGFGGGFGGGGGCCCCC	**AS:i:125**	**XS:i:102**	XN:i:0	XM:i:7	XO:i:0	XG:i:0	NM:i:7	MD:Z:10T0C9T1G27T7C3C14	YT:Z:UU	XC:Z:TTGGTTTAACTA	XM:Z:GCACCCTA	**AS:i:0**	**XS:i:0**	CY:Z:CCCCCGGGGGFG	UY:Z:GGFGCGGG	GE:Z:GENEA	XF:Z:CODING	GS:Z:+	gf:Z:CODING	gn:Z:GENEA	gs:Z:+

**Comment by jggatter** [comment]
The initial BAM had zero-value `AS` and `XS` tags that were being preserved  by `bowtie2 --preserve-tags`. The result was two pairs of AS and XS tags. My solution was to convert the tagged+trimmed BAM to SAM prior to alignment then use `gsed -e 's/\tAS\:i\:0\tXS\:i\://'` to remove these tags, then convert to a sorted BAM for alignment. After alignment, I faux-tagged genes and exons, this time using `sed -e 's@AS\:i\:[0-9]*XS\:i\:[0-9]*	@gf:Z:CODING	gn:Z:GENEA	gs:Z:+@'` (@jamesnemesh pointed out that I don't need those uppercase tags!)

Furthermore, the default tag for UMIs, `XM` was conflicting with bowtie2's `XM. I instead tagged UMIs with `TAG_NAME=XU` and fed `ADAPTER=^XU^XCACGTACTCTGCGTTGCTACCACTG`  to the PolyATrimmer and `MOLECULAR_BARCODE_TAG=XU` to DetectBeadSynthesisErrors and DetectBeadSubstitutionErrors. 

Strangely, it seems Picard's MergeBamAlignment is wiping some bowtie2 alignment tags or tag values (all XS are 0). I don't imagine I need these all however, so I'm just gonna take what I got and run with it for now.

Thanks so much @jamesnemesh, can't express enough how much I appreciated your time!
